README.md
# Project B***AI***ch

B**_AI_**ch is a project that generates songs using LSTM Neural Networks in Python using the Keras library.

The name is inspired on the great classical composer _Johann Sebastian Bach_.

All the songs used in the dataset are from [VGMusic](https://www.vgmusic.com/) and are creative common or public domain. It includes 100 songs used for training and validation.

## Contribute

If you want to contribute to our project, it's simple! We have a [Contribution Guide](docs/CONTRIBUTING.md) where all the steps to contribute are explained.
Don't forget to also read our [Code of Conduct](docs/CODE_OF_CONDUCT.md).
If you have any questions you can also contact us by creating an issue.

## Project Members

| Name | Registry |
| --- | --- |
| Isaque Alves | 14/0144544 |
| Mariana Mendes | 14/0154027 |
| João Pedro Sconetto | 14/0145940 |

## Requirements

* Python 3.x (Python 3.8 is recommended)
* Installing the following packages using pip:
	* Music21
	* Keras
	* Tensorflow
	* h5py

## External Dependencies

Some libraries are required for the base requirements to work properly. To install them run the following command (command for Debian based distro/linux):

```shell
sudo apt-get install portaudio19-dev python3-pyaudio
```

## Usage

### Training Model

To train the network you need to run the **lstm.py** file (make sure to have dependencies and your environment set as shown on `Development`).

```shell
python src/lstm.py
```

The network will use every MIDI file in `assets/songs` to train the network. The MIDI files should only contain a single instrument to get the most out of the training.

**NOTE**: You can stop the process at any point in time and the weights from the latest completed epoch will be available.

When you're satisfied with the loss of the AI select the weight you want to be use to predict and generate songs and rename it to `weights.hdf5` and leave it at `assets/weights`.

### Generating Music

Once you have trained the network you can generate text using **predict.py** file. Run the following command (make sure to have dependencies and your environment set as shown on `Development`).

```shell
python src/predict.py
```

### Known Errors

- music21 while reading mid files getting an decode error:

Check this correction: https://github.com/cuthbertLab/music21/pull/607/files

(intended for the next version but on version 6.1.0 this corrections is not present)

## Development

### Installing VirtualEnvWrapper

We recommend using a virtual environment created by the __virtualenvwrapper__ module. There is a virtual site with English instructions for installation that can be accessed [here](https://virtualenvwrapper.readthedocs.io/en/latest/install.html). But you can also follow these steps below for installing the environment:

```shell
sudo python3 -m pip install -U pip             # Update pip
sudo python3 -m pip install virtualenvwrapper  # Install virtualenvwrapper module
```

**Observation**: If you do not have administrator access on the machine remove `sudo` from the beginning of the command and add the flag `--user` to the end of the command.

Now configure your shell to use **virtualenvwrapper** by adding these two lines to your shell initialization file (e.g. `.bashrc`, `.profile`, etc.)

```shell
export WORKON_HOME=\$HOME/.virtualenvs
source /usr/local/bin/virtualenvwrapper.sh
```

If you want to add a specific project location (will automatically go to the project folder when the virtual environment is activated) just add a third line with the following `export`:

```shell
export PROJECT_HOME=/path/to/project
```

Run the shell startup file for the changes to take effect, for example:

```shell
source ~/.bashrc
```

Now create a virtual environment with the following command (entering the name you want for the environment), in this example I will use the name **baich**:

```shell
mkvirtualenv -p $(which python3) baich
```

To use it:

```shell
workon baich
sudo python3 -m pip install pipenv
pipenv install # Will install all of the project dependencies
# or
pip install -r requirements.txt
```

**Observaion**: Again, if necessary, add the flag `--user` to make the pipenv package installation for the local user.

### Local Execution

For local system execution, run the following command in the project root folder (assuming virtualenv is already active):

```shell
python src/lstm.py
```

This will train and run the system on your machine. This way you can test new implementations or new optmizations. Also you can create songs for testing.

### Test

#### Lint

To lint your code follow the script bellow:

1. Enable virtualenv _baich_;

2. Ensure that the dependencies are installed, especially:

```code
black
flake8
```

3. Run the command below:

```shell
black src/*
flake8 --ignore E501 src
```

During the lint process the terminal will report a code errors and warnings from the PEP8 style guide, for more configurations and additional documentation go to [flake8](http://flake8.pycqa.org/en/latest/) and [PEP8](https://www.python.org/dev/peps/pep-0008/).
The rule `E501` (Line too long) is ignored because the base linter is black and not PEP8.

## Build

### Generate Changelog

To generate changelog we use the `standard version` tool, it will auto generate a new changelog for every new release by using the commit messages. To generate a new release and generate the updated changelog just do the following steps:

1. Install all dependencies

```shell
yarn install
```

2. Run standard version:

```shell
yarn run release
```

If the release is a pre-release you should add the `--prerelease` to the command:

```shell
yarn run release -- --prerelease alpha
```

For further instructions or other options check the full documentation of `standard version` project in the [CLI Usage](https://github.com/conventional-changelog/standard-version#cli-usage) section.

## References

Skúli, Sigurður (Dec, 2017). How to Generate Music using a LSTM Neural Network in Keras. Towards Data Science. Access date: 10 Nov 2020. Available at <https://towardsdatascience.com/how-to-generate-music-using-a-lstm-neural-network-in-keras-68786834d4c5>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 0.1.0 (2020-11-21)


### Features

* add asset files ([13a9040](https://github.com/deeplearningunb/baich/commit/13a9040e7f1bab03f63637cbe27356a2d3788afc))
* **ai:** create base version ([38d92e8](https://github.com/deeplearningunb/baich/commit/38d92e847adda332e3a8e55562148334b09aa920))
* **assets:** add base songs for training ([732d680](https://github.com/deeplearningunb/baich/commit/732d680fd193345febce88f337a7eb587e8e42f5))
* **deps:** add child dependencies ([80264b3](https://github.com/deeplearningunb/baich/commit/80264b300507d3d686be495ed2038f9015a209ae))
* **deps:** add project's dependencies ([4ae08bd](https://github.com/deeplearningunb/baich/commit/4ae08bd7ee60e7b635c2dde9c0238114b378ba6b))
* **deps:** update dependency version ([1f1f948](https://github.com/deeplearningunb/baich/commit/1f1f948b2a3ebe05960a8dbc3330f0ee5a0afbc4))
* **deps:** update requirements file ([410c977](https://github.com/deeplearningunb/baich/commit/410c9777a146c610bd35b09936f33b1510adff0e))
* **docs:** add external dependencies ([f0dd699](https://github.com/deeplearningunb/baich/commit/f0dd69901e022a5738516ddb8ad8ed42c5ac17ac))
* **docs:** add known issues for the project ([61c156f](https://github.com/deeplearningunb/baich/commit/61c156fa14bbf14c0b8dfece724f25ea1901657d))
* **docs:** evolve project documentation ([03688b9](https://github.com/deeplearningunb/baich/commit/03688b9cc6bc7daa8a8b1a8f8a49e0e089eb3505))
* **docs:** last updates before release ([0b8e08e](https://github.com/deeplearningunb/baich/commit/0b8e08e9d5faa428a01e4631600aa24074711c76))
* **notes:** add notes binary file ([935aa52](https://github.com/deeplearningunb/baich/commit/935aa5205117a9c7401e67ac67a26e454a25b150))
* **src:** insert logs ([ebf8619](https://github.com/deeplearningunb/baich/commit/ebf86198a229ba8564dd88c1ba412e13e8fcc86f))
* initial commit ([5908218](https://github.com/deeplearningunb/baich/commit/590821859f5cb5da77b91df234b1d8f33a5b8376))


### Bug Fixes

* **deps:** minor name fix ([f2ebd54](https://github.com/deeplearningunb/baich/commit/f2ebd54eb75e1e514e4f6426fa77a2c62d0594c1))
* **docs:** update docs ([b68fb3e](https://github.com/deeplearningunb/baich/commit/b68fb3e468ae19815330dc3a255a125fa6186b8b))
* **docs:** Update docs ([bed398d](https://github.com/deeplearningunb/baich/commit/bed398d7fabb8f99b145ca57779893661c4e686b))
* **src:** create basic logging configuration ([073ad73](https://github.com/deeplearningunb/baich/commit/073ad734e4f6a6e18391259eb7452662ac3f3c2b))


### Refactors

* **ai:** apply black and correct lint errors ([14e23ea](https://github.com/deeplearningunb/baich/commit/14e23ea9201ba7fc1c08bd103ac8091f3440f69f))
* **predict:** apply black ([49f9418](https://github.com/deeplearningunb/baich/commit/49f941842adc1087a355ebf468deb428076d945e))


### Chore

* add test music from 21/11/2020 ([71a6e0c](https://github.com/deeplearningunb/baich/commit/71a6e0c0fabd3a11403930f79cc4727631183eab))
* **assets:** reupdate midi files ([fea8fc4](https://github.com/deeplearningunb/baich/commit/fea8fc473572147381fbef7bdb5e1459624f9c43))
* create project gitignore ([cf5ec45](https://github.com/deeplearningunb/baich/commit/cf5ec45925718db2ea27ef3a315f5763cdf9e625))
* update gitignore ([9c10d71](https://github.com/deeplearningunb/baich/commit/9c10d71e3b3305243701e7e3968dd679696881af))


src/lstm.py
""" This module prepares midi file data and feeds it to the neural
    network for training """
import glob
import pickle
import numpy
import logging
from music21 import converter, instrument, note, chord
from keras.models import Sequential
from keras.layers import Dense, Dropout, LSTM, Activation
from keras.layers import BatchNormalization as BatchNorm
from keras.utils import np_utils
from keras.callbacks import ModelCheckpoint


logging.basicConfig(
    level=logging.DEBUG, format="%(asctime)s [%(levelname)s] - %(message)s"
)
_logger = logging.getLogger(__name__)


def train_network():
    """ Train a Neural Network to generate music """
    logging.info("Start Train Network")
    notes = get_notes()

    # get amount of pitch names
    n_vocab = len(set(notes))
    logging.info(f"Amount of pitches:{n_vocab}")

    network_input, network_output = prepare_sequences(notes, n_vocab)

    model = create_network(network_input, n_vocab)

    train(model, network_input, network_output)


def get_notes():
    """ Get all the notes and chords from the midi files in the ./assets/songs directory """
    logging.info("Start get all notes and chords")
    notes = []

    for file in glob.glob("assets/songs/*.mid"):
        midi = converter.parse(file)

        _logger.info(f"Parsing song: {file}")

        notes_to_parse = None

        # file has instrument parts
        try:
            s2 = instrument.partitionByInstrument(midi)
            notes_to_parse = s2.parts[0].recurse()
        # file has notes in a flat structure
        except Exception:
            notes_to_parse = midi.flat.notes
            logging.error("Exception occurred", exc_info=True)

        for element in notes_to_parse:
            if isinstance(element, note.Note):
                notes.append(str(element.pitch))
            elif isinstance(element, chord.Chord):
                notes.append(".".join(str(n) for n in element.normalOrder))

    with open("data/notes", "wb") as filepath:
        pickle.dump(notes, filepath)
        logging.info("written notes in data/notes")
    return notes


def prepare_sequences(notes, n_vocab):
    """ Prepare the sequences used by the Neural Network """
    logging.info("Starting to prepar the sequence")
    sequence_length = 100

    # get all pitch names
    pitchnames = sorted(set(item for item in notes))

    # create a dictionary to map pitches to integers
    note_to_int = dict((note, number) for number, note in enumerate(pitchnames))
    logging.info("Dictionary created")

    network_input = []
    network_output = []

    # create input sequences and the corresponding outputs
    for i in range(0, len(notes) - sequence_length, 1):
        sequence_in = notes[i:i + sequence_length]
        sequence_out = notes[i + sequence_length]
        network_input.append([note_to_int[char] for char in sequence_in])
        network_output.append(note_to_int[sequence_out])

    _logger.info("Network input and output created")
    n_patterns = len(network_input)
    logging.info("Number of patterns:%s", n_patterns)

    # reshape the input into a format compatible with LSTM layers
    network_input = numpy.reshape(network_input, (n_patterns, sequence_length, 1))
    # normalize input
    network_input = network_input / float(n_vocab)
    _logger.info("Network input reshaped and normalized")

    network_output = np_utils.to_categorical(network_output)
    logging.info("Sequences Prepared")
    return (network_input, network_output)


def create_network(network_input, n_vocab):
    """ create the structure of the neural network """
    logging.info("Starting to create network")
    model = Sequential()
    model.add(
        LSTM(
            512,
            input_shape=(network_input.shape[1], network_input.shape[2]),
            recurrent_dropout=0.3,
            return_sequences=True,
        )
    )
    model.add(
        LSTM(
            512,
            return_sequences=True,
            recurrent_dropout=0.3,
        )
    )
    model.add(LSTM(512))
    model.add(BatchNorm())
    model.add(Dropout(0.3))
    model.add(Dense(256))
    model.add(Activation("relu"))
    model.add(BatchNorm())
    model.add(Dropout(0.3))
    model.add(Dense(n_vocab))
    model.add(Activation("softmax"))
    model.compile(loss="categorical_crossentropy", optimizer="rmsprop")

    logging.info("Neural Network created")
    return model


def train(model, network_input, network_output):
    """ train the neural network """
    logging.info("Starting model training")
    filepath = "assets/weights/weights-improvement-{epoch:02d}-{loss:.4f}-bigger.hdf5"
    checkpoint = ModelCheckpoint(
        filepath, monitor="loss", verbose=0, save_best_only=True, mode="min"
    )
    callbacks_list = [checkpoint]

    model.fit(
        network_input,
        network_output,
        epochs=200,
        batch_size=128,
        callbacks=callbacks_list,
    )
    logging.info("Model Trained")


if __name__ == "__main__":
    train_network()


src/predict.py
""" This module generates notes for a midi file using the
    trained neural network """
import pickle
import numpy
import logging
from datetime import datetime
from music21 import instrument, note, stream, chord
from keras.models import Sequential
from keras.layers import Dense, Dropout, LSTM, Activation
from keras.layers import BatchNormalization as BatchNorm

logging.basicConfig(
    level=logging.DEBUG, format="%(asctime)s [%(levelname)s] - %(message)s"
)
_logger = logging.getLogger(__name__)


def generate():
    """ Generate a piano midi file """
    # load the notes used to train the model
    with open("data/notes", "rb") as filepath:
        notes = pickle.load(filepath)
        _logger.info("Notes and Acords Loaded")

    # Get all pitch names
    pitchnames = sorted(set(item for item in notes))
    # Get amount of pitch names
    n_vocab = len(set(notes))
    _logger.info(f"Getting all {n_vocab} pitch names")

    network_input, normalized_input = prepare_sequences(notes, pitchnames, n_vocab)
    model = create_network(normalized_input, n_vocab)
    prediction_output = generate_notes(model, network_input, pitchnames, n_vocab)
    create_midi(prediction_output)
    _logger.info("MIDI song generated")


def prepare_sequences(notes, pitchnames, n_vocab):
    """ Prepare the sequences used by the Neural Network """
    # map between notes and integers and back
    _logger.info("Starting to prepar the sequence")
    note_to_int = dict((note, number) for number, note in enumerate(pitchnames))

    sequence_length = 100
    network_input = []
    output = []
    for i in range(0, len(notes) - sequence_length, 1):
        sequence_in = notes[i:i + sequence_length]
        sequence_out = notes[i + sequence_length]
        network_input.append([note_to_int[char] for char in sequence_in])
        output.append(note_to_int[sequence_out])

    n_patterns = len(network_input)
    _logger.info(f"Network input and output created with {n_patterns} patterns")
    # reshape the input into a format compatible with LSTM layers
    normalized_input = numpy.reshape(network_input, (n_patterns, sequence_length, 1))
    # normalize input
    normalized_input = normalized_input / float(n_vocab)

    _logger.info("Network input reshaped and normalized")

    return (network_input, normalized_input)


def create_network(network_input, n_vocab):
    """ create the structure of the neural network """
    _logger.info("Starting to create network")
    model = Sequential()
    model.add(
        LSTM(
            512,
            input_shape=(network_input.shape[1], network_input.shape[2]),
            recurrent_dropout=0.3,
            return_sequences=True,
        )
    )
    model.add(
        LSTM(
            512,
            return_sequences=True,
            recurrent_dropout=0.3,
        )
    )
    model.add(LSTM(512))
    model.add(BatchNorm())
    model.add(Dropout(0.3))
    model.add(Dense(256))
    model.add(Activation("relu"))
    model.add(BatchNorm())
    model.add(Dropout(0.3))
    model.add(Dense(n_vocab))
    model.add(Activation("softmax"))
    model.compile(loss="categorical_crossentropy", optimizer="rmsprop")

    # Load the weights to each node
    _logger.info("Loading weights")
    model.load_weights("assets/weights/weights.hdf5")
    _logger.info("Neural Network created")
    return model


def generate_notes(model, network_input, pitchnames, n_vocab):
    """ Generate notes from the neural network based on a sequence of notes """
    # pick a random sequence from the input as a starting point for the prediction
    _logger.info("Starting generate notes")
    start = numpy.random.randint(0, len(network_input) - 1)

    int_to_note = dict((number, note) for number, note in enumerate(pitchnames))

    pattern = network_input[start]
    prediction_output = []

    # generate 500 notes
    _logger.info("Starting the prediction")
    for note_index in range(500):
        prediction_input = numpy.reshape(pattern, (1, len(pattern), 1))
        prediction_input = prediction_input / float(n_vocab)

        prediction = model.predict(prediction_input, verbose=0)

        index = numpy.argmax(prediction)
        result = int_to_note[index]
        _logger.info(f"Prediction result: {result}")
        prediction_output.append(result)

        pattern.append(index)
        pattern = pattern[1:len(pattern)]
    _logger.info("Prediction ending")
    return prediction_output


def create_midi(prediction_output):
    """convert the output from the prediction to notes and create a midi file
    from the notes"""
    offset = 0
    output_notes = []

    # create note and chord objects based on the values generated by the model
    _logger.info("Creating notes and acords")
    for pattern in prediction_output:
        # pattern is a chord
        if ("." in pattern) or pattern.isdigit():
            notes_in_chord = pattern.split(".")
            notes = []
            for current_note in notes_in_chord:
                new_note = note.Note(int(current_note))
                new_note.storedInstrument = instrument.Piano()
                notes.append(new_note)
            new_chord = chord.Chord(notes)
            new_chord.offset = offset
            output_notes.append(new_chord)
        # pattern is a note
        else:
            new_note = note.Note(pattern)
            new_note.offset = offset
            new_note.storedInstrument = instrument.Piano()
            output_notes.append(new_note)

        # increase offset each iteration so that notes do not stack
        offset += 0.5

    midi_stream = stream.Stream(output_notes)

    midi_stream.write(
        "midi", fp=f"assets/composed/{datetime.now().strftime('%Y.%m.%d_%H.%M_')}test_output.mid"
    )
    _logger.info("Music created!")


if __name__ == "__main__":
    _logger.info("Starting the Prediction")
    generate()
